File: python/tools/file_tools.py
import base64
import json
import os
from langchain_core.tools import tool


@tool
def read_file(file_path: str) -> str:
    """读取文件内容。参数 file_path: 文件的完整路径"""
    try:
        with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
            content = f.read()
        if len(content) > 5000:
            content = content[:5000] + "\n...(内容过长,已截断)"
        return content
    except FileNotFoundError:
        return f"文件不存在: {file_path}"
    except PermissionError:
        return f"没有权限读取文件: {file_path}"
    except Exception as e:
        return f"读取文件出错: {str(e)}"


@tool
def write_file(file_path: str, content: str) -> str:
    """写入内容到文件。参数 file_path: 文件的完整路径, content: 要写入的内容"""
    try:
        os.makedirs(os.path.dirname(file_path) or '.', exist_ok=True)
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        return f"文件已写入: {file_path} ({len(content)} 字符)"
    except PermissionError:
        return f"没有权限写入文件: {file_path}"
    except Exception as e:
        return f"写入文件出错: {str(e)}"


@tool
def read_image(file_path: str) -> str:
    """读取图片文件,获取图片信息。参数 file_path: 图片文件的完整路径。
    支持的格式: PNG, JPG, JPEG, GIF, BMP, WEBP, SVG
    返回图片的 base64 编码和前 200 字符预览,供视觉模型分析。
    SVG 会直接返回文本内容。"""
    import base64

    if not os.path.exists(file_path):
        return f"图片不存在: {file_path}"

    ext = os.path.splitext(file_path)[1].lower()

    # SVG 直接读文本
    if ext == '.svg':
        try:
            with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
                svg_content = f.read(3000)
            return f"[SVG图片] {file_path}\n内容:\n{svg_content}"
        except Exception as e:
            return f"读取 SVG 出错: {str(e)}"

    # 普通图片: base64 编码
    try:
        file_size = os.path.getsize(file_path)
        if file_size > 5 * 1024 * 1024:
            return f"图片过大({file_size / 1024 / 1024:.1f}MB),请压缩后再分析"

        mime_map = {'.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
                    '.gif': 'image/gif', '.bmp': 'image/bmp', '.webp': 'image/webp'}
        mime = mime_map.get(ext, 'image/png')

        with open(file_path, 'rb') as f:
            raw = f.read()
        b64 = base64.b64encode(raw).decode('ascii')
        data_url = f"data:{mime};base64,{b64}"

        return json.dumps({
            "type": "image_analysis_request",
            "file_path": file_path,
            "file_size": file_size,
            "mime_type": mime,
            "data_url": data_url[:100] + "...(base64数据, 请在下一轮对话中引用此图片)",
            "instruction": "请告知用户需要通过粘贴或拖拽图片到对话框来让AI直接分析图片内容。或者使用 read_image 工具返回的 base64 数据。",
        }, ensure_ascii=False)
    except Exception as e:
        return f"读取图片出错: {str(e)}"


@tool
def list_files(directory: str = ".") -> str:
    """列出目录中的文件。参数 directory: 目录路径,默认为当前目录"""
    try:
        items = os.listdir(directory)
        result = []
        for item in sorted(items):
            full_path = os.path.join(directory, item)
            tag = "[DIR]" if os.path.isdir(full_path) else "[FILE]"
            try:
                size = os.path.getsize(full_path)
                if size > 1024 * 1024:
                    size_str = f"{size / (1024*1024):.1f}MB"
                elif size > 1024:
                    size_str = f"{size / 1024:.1f}KB"
                else:
                    size_str = f"{size}B"
            except:
                size_str = "?"
            result.append(f"  {tag} {item} ({size_str})")
        return f"目录 {directory}:\n" + "\n".join(result[:50])
    except FileNotFoundError:
        return f"目录不存在: {directory}"
    except Exception as e:
        return f"列出目录出错: {str(e)}"

File: server/lib/tools/file-tools.js
/**
 * file-tools.js — 文件操作工具
 *
 * read_file, write_file, list_files, read_image
 */
import fs from 'node:fs';
import path from 'node:path';

const MAX_READ_SIZE = 100_000; // 100KB max read
const MAX_WRITE_SIZE = 500_000; // 500KB max write
const DANGEROUS_EXTENSIONS = new Set(['.exe', '.dll', '.sys', '.com', '.bat', '.cmd', '.ps1', '.vbs']);

function _safePath(filepath) {
  // Resolve and normalize
  const resolved = path.resolve(filepath);
  // Block writing to dangerous extensions via read_image
  return resolved;
}

export const readFile = {
  name: 'read_file',
  description: '读取文件内容。参数 file_path: 文件路径（绝对路径）',
  parameters: {
    type: 'object',
    properties: {
      file_path: { type: 'string', description: '要读取的文件绝对路径' },
    },
    required: ['file_path'],
  },
  async invoke({ file_path }) {
    try {
      const resolved = _safePath(file_path);
      if (!fs.existsSync(resolved)) return `文件不存在: ${file_path}`;
      const stat = fs.statSync(resolved);
      if (stat.isDirectory()) return `这是一个目录，不是文件: ${file_path}`;
      if (stat.size > MAX_READ_SIZE) return `文件过大 (${(stat.size / 1024).toFixed(0)}KB)，超过 ${MAX_READ_SIZE / 1024}KB 限制`;
      const content = fs.readFileSync(resolved, 'utf-8');
      return content;
    } catch (err) {
      return `文件读取失败: ${err.message}`;
    }
  },
};

export const writeFile = {
  name: 'write_file',
  description: '写入文件内容。参数 file_path: 文件路径，content: 要写入的内容',
  parameters: {
    type: 'object',
    properties: {
      file_path: { type: 'string', description: '要写入的文件绝对路径' },
      content: { type: 'string', description: '要写入的内容' },
    },
    required: ['file_path', 'content'],
  },
  async invoke({ file_path, content }) {
    try {
      if (!content) return '内容为空，拒绝写入';
      if (content.length > MAX_WRITE_SIZE) return `内容过大 (${content.length} 字符)，超过 ${MAX_WRITE_SIZE} 字符限制`;
      const resolved = _safePath(file_path);
      const dir = path.dirname(resolved);
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
      fs.writeFileSync(resolved, content, 'utf-8');
      return `文件已写入: ${file_path} (${content.length} 字符)`;
    } catch (err) {
      return `文件写入失败: ${err.message}`;
    }
  },
};

export const listFiles = {
  name: 'list_files',
  description: '列出目录下的文件。参数 dir_path: 目录路径',
  parameters: {
    type: 'object',
    properties: {
      dir_path: { type: 'string', description: '要列出的目录绝对路径' },
    },
    required: ['dir_path'],
  },
  async invoke({ dir_path }) {
    try {
      const resolved = _safePath(dir_path);
      if (!fs.existsSync(resolved)) return `目录不存在: ${dir_path}`;
      if (!fs.statSync(resolved).isDirectory()) return `这不是一个目录: ${dir_path}`;
      const entries = fs.readdirSync(resolved, { withFileTypes: true });
      const lines = entries.slice(0, 100).map(e => {
        const type = e.isDirectory() ? 'DIR' : e.isFile() ? 'FILE' : 'OTHER';
        const size = e.isFile() ? ` (${(fs.statSync(path.join(resolved, e.name)).size / 1024).toFixed(1)}KB)` : '';
        return `[${type}] ${e.name}${size}`;
      });
      if (entries.length > 100) lines.push(`... 还有 ${entries.length - 100} 个条目`);
      return lines.join('\n');
    } catch (err) {
      return `列出目录失败: ${err.message}`;
    }
  },
};

export const readImage = {
  name: 'read_image',
  description: '读取图片文件，返回 base64 data URL。参数 file_path: 图片文件路径',
  parameters: {
    type: 'object',
    properties: {
      file_path: { type: 'string', description: '要读取的图片文件绝对路径' },
    },
    required: ['file_path'],
  },
  async invoke({ file_path }) {
    try {
      const resolved = _safePath(file_path);
      if (!fs.existsSync(resolved)) return `文件不存在: ${file_path}`;
      const ext = path.extname(resolved).toLowerCase();
      const mimeMap = { '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg', '.gif': 'image/gif', '.webp': 'image/webp', '.bmp': 'image/bmp' };
      const mime = mimeMap[ext] || 'application/octet-stream';
      const buf = fs.readFileSync(resolved);
      const b64 = buf.toString('base64');
      return `data:${mime};base64,${b64}`;
    } catch (err) {
      return `图片读取失败: ${err.message}`;
    }
  },
};
